File: DP_Problems/gridTraveller/gridTraveler.js
const gridTraveler = (x, y, memo = {}) => {
  const key = x + ',' + y
  if (key in memo) return memo[key]
  if (x == 1 || y == 1) return 1
  if (x == 0 || y == 0) return 0
  memo[key] = gridTraveler(x - 1, y, memo) + gridTraveler(x, y - 1, memo)
  return memo[key]
}

console.log(gridTraveler(1, 1))
console.log(gridTraveler(2, 3))
console.log(gridTraveler(3, 2))
console.log(gridTraveler(3, 3))
console.log(gridTraveler(18, 18))

File: DP_Problems/gridTraveller/gridTraveler.cpp
#include <iostream>
#include <ctime>
#include <cstdlib>
#include <map>
#include <utility>

using namespace std;

long int gridTraveler(int x, int y)
{

    if (x == 1 || y == 1)
        return 1;
    if (x == 0 || y == 0)
        return 0;

    return gridTraveler(x - 1, y) + gridTraveler(x, y - 1);
}

map<pair<int, int>, long int> M;
long int memoGridTraveler(int x, int y)
{
    pair<int, int> p(x, y);
    if (M[p] != 0)
        return M[p];
    if (x == 1 || y == 1)
        return 1;
    if (x == 0 || y == 0)
        return 0;
    M[p] = memoGridTraveler(x - 1, y) + memoGridTraveler(x, y - 1);
    return M[p];
}

int test(int (*func)(int, int))
{
    for (int i = 0; i < 10; i++)
    {
        int randX = (rand() % 20) + 1;
        int randY = (rand() % 20) + 1;
        cout << "Input " << randX << ", " << randY << ", output: " << func(randX, randY) << endl;
    }
}

int main()
{
    // test(&gridTraveler);
    // cout << "**** Memoized Solution ****" << endl;
    // test(&memoGridTraveler);

    cout << memoGridTraveler(1, 1) << endl;
    cout << memoGridTraveler(2, 3) << endl;
    cout << memoGridTraveler(3, 2) << endl;
    cout << memoGridTraveler(3, 3) << endl;
    cout << memoGridTraveler(18, 18) << endl;

    return 0;
}
